doc/user/migrating-existing-apis.md

```diff
@@ -0,0 +1,384 @@
+
+- Changing the package statements in each file: `package v1` -> `package shared`.
+- Exporting types, their methods, and functions used by external API types.
+
+This will regenerate deepcopy code for all tagged types in `pkg/apis`.
+
+Additionally, `deepcopy-gen` must be run on the new package to generate `DeepCopy()` and `DeepCopyInto()` methods, which are necessary for all Kubernetes API code. Doing so will also remove deepcopy code for the now `shared` types from `v1`. To do so, run the following command:
+
+```console
+$ operator-sdk generate k8s
+```
+
+Now that shared types and functions have their own package we can update types in package `v1`, and any other package that imports them from `v1`, to use those in `shared`. The source file `catalogsourceconfig_types.go` imports and uses a type defined in `shared`, `ObjectPhase`, as a field in `CatalogSourceConfigSpec`.
+
+```Go
+import (
+  "github.com/test-org/test-operator/pkg/apis/operators/shared"
+
+  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+...
+
+type CatalogSourceConfig struct {
+  metav1.TypeMeta `json:",inline"`
+  metav1.ListMeta `json:"metadata,omitempty"`
+
+  Spec              CatalogSourceConfigSpec   `json:"spec,omitempty"`
+  Status            CatalogSourceConfigStatus `json:"status,omitempty"`
+}
+
+type CatalogSourceConfigSpec struct {
+  ...
+}
+
+type CatalogSourceConfigStatus struct {
+  // The type was previously ObjectPhase, now shared.ObjectPhase.
+  CurrentPhase shared.ObjectPhase `json:"currentPhase,omitempty"`
+  ...
+}
+```
+
+Do this for all instances of types previously in `v1` that are now in `shared`. Once done, remove `phase.go`, `phase_types.go`, and `shared.go` from `pkg/apis/operators/v1`.
+
+### Updating empty v2 types using v1 types
+
+The `CatalogSourceConfig` type and schema code were generated by `operator-sdk add api`, but the types are not populated. We need to move existing type data from `v1` to `v2`. This process is similar to migrating shared code, except we do not need to export any types or functions.
+
+Remove `pkg/apis/operators/v2/catalogsourceconfig_types.go` and copy `catalogsourceconfig.go` and `catalogsourceconfig_types.go` from `pkg/apis/operators/v1` to `pkg/apis/operators/v2`:
+
+```console
+$ rm pkg/apis/operators/v2/catalogsourceconfig_types.go
+$ cp pkg/apis/operators/v1/catalogsourceconfig*.go pkg/apis/operators/v2
+```
+
+If you have any comments or custom code in `pkg/apis/operators/v1` related to source code in either copied file, ensure that is copied to `doc.go` or `register.go` in `pkg/apis/operators/v2`.
+
+You can now run `operator-sdk generate k8s` to generate deepcopy code for the migrated `v2` types. Once this is done, update all packages that import the migrated `v1` types to use those in `v2`.
+
+The final step is to remove `catalogsourceconfig.go` and `catalogsourceconfig_types.go` and any related comments or custom code in `doc.go` or `register.go` from `pkg/apis/operators/v1`.
```

Comment:
We mention backwards-compatibility earlier in the document. Should we discuss it more here? In general, I think the default should be to keep the old types around and document them as deprecated, especially for non-alpha and non-beta versions and versions that have already been released to users.

Reply:
Following k8s versioning conventions and what you suggest, the developer should create a new `shared` and `v2` package, but keep all the code in `v1` as-is (with deprecation comments). I think that's a reasonable recommendation here as well.